BraveWallet/Crypto/Stores/PortfolioStore.swift

```diff
@@ -94,137 +98,142 @@ public class PortfolioStore: ObservableObject {
-  private func fetchBalances(for tokens: [BraveWallet.BlockchainToken], accounts: [BraveWallet.AccountInfo], completion: @escaping ([String: Double]) -> Void) {
-    var balances: [String: Double] = [:]
-    let group = DispatchGroup()
-    for account in accounts {
-      for token in tokens {
-        group.enter()
-        rpcService.balance(for: token, in: account) { balance in
-          defer { group.leave() }
-          guard let balance = balance else {
-            return
+  
+  func update() {
+    self.updateTask?.cancel()
+    self.updateTask = Task { @MainActor in
+      self.isLoadingBalances = true
+      let coin = await walletService.selectedCoin()
+      let network = await rpcService.network(coin)
+      let userAssets = await walletService.userAssets(network.chainId, coin: coin)
+      // if the task was cancelled, don't update the UI
+      guard !Task.isCancelled else { return }
+      // update userVisibleAssets on display immediately with empty values. Issue #5567
+      userVisibleAssets = userAssets.map { token in
+        AssetViewModel(
+          token: token,
+          decimalBalance: 0.0,
+          price: "",
+          history: []
+        )
+      }
+      
+      let keyring = await keyringService.keyringInfo(coin.keyringId)
+      guard !Task.isCancelled else { return } // limit network request(s) if cancelled
+      let balances = await fetchBalances(for: userAssets, accounts: keyring.accountInfos)
+      guard !Task.isCancelled else { return } // limit network request(s) if cancelled
+      let nonZeroBalanceTokens = balances.filter { $1 > 0 }.map { $0.key }
+      let priceHistories = await fetchPriceHistory(for: nonZeroBalanceTokens)
+      guard !Task.isCancelled else { return } // limit network request(s) if cancelled
+      let visibleTokenSymbols = userAssets.filter(\.visible).map { $0.symbol.lowercased() }
+      let prices = await fetchPrices(for: visibleTokenSymbols)
+      
+      // if the task was cancelled, don't update the UI
+      guard !Task.isCancelled else { return }
+      
+      // build our userVisibleAssets
+      userVisibleAssets = userAssets.map { token in
+        let symbol = token.symbol.lowercased()
+        return AssetViewModel(
+          token: token,
+          decimalBalance: balances[symbol] ?? 0.0,
+          price: prices[symbol] ?? "",
+          history: priceHistories[symbol] ?? []
+        )
+      }
+      // Compute balance based on current prices
+      let currentBalance = userVisibleAssets
+        .compactMap {
+          if let price = Double($0.price) {
+            return $0.decimalBalance * price
           }
-          let symbol = token.symbol.lowercased()
-          balances[symbol, default: 0] += balance
+          return nil
         }
+        .reduce(0.0, +)
+      balance = currencyFormatter.string(from: NSNumber(value: currentBalance)) ?? "–"
```

Comment:
why its a `-` lol

Reply:
Existing code, I assume `-` instead of `0` to differentiate between zero balance and failed to load/parse balance.